### DISCUSSION TITLE
Sitemap index

### DISCUSSION THREAD

**Discussion by omerp-explorium** [vercel/next.js#61025] [Ideas]
### Goals

When using the multiple sitmap approach, Nextjs should produce a root sitemap that will point to all sitemap urlsets e.g <domain>/sitemap.xml

### Non-Goals

1.
2.
3.

### Background

As you can see in the image, we should also have this sitemap pointing to multiple sitemaps
![image](https://github.com/vercel/next.js/assets/93073822/3c648809-d386-4ff1-9d66-81d82bd175a0)

Taken from [Google documentation on large sitemaps](https://developers.google.com/search/docs/crawling-indexing/sitemaps/large-sitemaps)

### Proposal

I propose Nextjs will keep track of all the generated sitemaps, and generate the proposed sitemap of sitemaps xml.

What do you think?

**Comment by Fedeorlandau** [comment]
I made a PR that adds support for `sitemap-index.ts` which you can use to dynamically add your sitemaps to the sitemap index file.

https://github.com/vercel/next.js/pull/61391

**Reply by ricardo-fnd** [reply]
@Fedeorlandau But this means that we do not need `/app/sitemap.xml` working? Or `sitemap-index.ts` is the `sitemap.xml` which points for others sitemaps?

**Reply by Fedeorlandau** [reply]
Hi! This is just for having a sitemap-index file where you can manually generate the index and put your other sitemaps files. You still need to generate sitemap.xml files for your site.

**Reply by tomcru** [reply]
Thanks for the PR @Fedeorlandau. Hope to see this merged soon!

Could you possible give any pointers on how you have handled sitemap index creation until now?

**Reply by Fedeorlandau** [reply]
@tomcru absolutely. The way I found it to work is by creating a /app/sitemap-index.xml/route.ts with a route GET handler and returning a xml like response.

**Reply by tomcru** [reply]
For anyone else coming across this, here's a quick example I used to test a sitemap index via `/app/sitemap-index.xml/route.ts` as Fedeorlandau suggested.

```typescript
const PAGE_SIZE = 1000;

const generateSitemapLink = (url: string) =>
  `<sitemap><loc>${url}</loc></sitemap>`;
  
export async function GET() {
  const gamesResponse = await fetch(
    `${API.base}/products?${new URLSearchParams({
      pageSize: String(PAGE_SIZE),
      pageNumber: String(1),
      currency: Currency.usd,
      order: 'nameAsc',
    })}`,
  );

  const games = (await gamesResponse.json()) as GetProductsOkResponse;

  const sitemapIndexXML = `<?xml version="1.0" encoding="UTF-8"?>
    <sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
        ${generateSitemapLink(`${URLS.base}/sitemap.xml`)}

        ${Array.from({ length: games.data.pageCount }, (_, i) => i + 1)
          .map((id) =>
            generateSitemapLink(`${URLS.base}/games/sitemap/${id}.xml`),
          )
          .join('')} 
    </sitemapindex>`;

  return new Response(sitemapIndexXML, {
    headers: { 'Content-Type': 'text/xml' },
  });
}
```

Result:
<img width="950" alt="Screenshot 2024-02-13 at 02 09 08" src="https://github.com/vercel/next.js/assets/35841182/36cd9149-0f94-40f2-a88c-5fa8e3888d87">

**Reply by MartineBoulanger** [reply]
Thank you for this! I was looking for a solution since this sitemapIndex() is still a WIP, and your solution works perfectly!

**Reply by rkuang9** [reply]
Wouldn't sitemaps generated this way be ignored by crawlers? If it's `/games/sitemap/1.xml`, the sitemap may only point to link within its current level, which is `/games/sitemap`.

**Reply by tomcru** [reply]
Not sure I understand your question @rkuang9 
The sitemaps aren't ignored by crawlers and all properly processed by Google, etc.

Our link structure is like this atm and the sitemap index is on `/sitemaps/index.xml`
<img width="559" alt="Screenshot 2025-04-07 at 11 16 53" src="https://github.com/user-attachments/assets/44570e49-a1ff-460e-b214-83eb2610e3dc" />

**Reply by rkuang9** [reply]
It's a matter of directory levels. According to https://developers.google.com/search/docs/crawling-indexing/sitemaps/large-sitemaps#sitemap-index-best-practices and https://www.sitemaps.org/protocol.html#location, your links in `sitemaps/games/1.xml` will only be recognized if they point to `sitemaps/games`. Of course this directory is dedicated to sitemaps and not game content. It seems the way Next.js handles sitemaps do not fit sitemap standards.

**Reply by tomcru** [reply]
The first article linked is only talking about sitemap indexes, which is correct in our case as all sitemap indexes are in the `/sitemaps/` directory. It's saying, however, that you shouldn't have your index at `/sitemaps/index.xml` and then link to another sitemap that's in another directory like `/somewhere-else/sitemap.xml`.

I'm also not sure if link 2 is enforced, I've never run into any problem. It would mean I could only link to game pages that are in `/sitemaps/games/...`. But this is not the case as Google is clearly recognising the sitemaps and their included urls (on `/games/...`)

Also this has nothing to do with next.js, as you can choose the directory/path you want your sitemaps or indexes to be yourself.

**Comment by antoinerousseau** [comment]
I don't understand the utility of `generateSitemaps` without a sitemaps index, am I missing something?

**Reply by hermbo** [reply]
Agreed, seems like this should just be automatic. Then you only need to provide google with the index file.

**Reply by CaptainN** [reply]
Next.js has been filling up with half baked "features" - I'm not sure why...

**Reply by shadee22** [reply]
Yes, we also need a feature to generate a NEXTJS sitemap index. Right now, it only generates URLs.

**Comment by antoinerousseau** [comment]
Meanwhile, I went with my own custom index at `src/app/sitemap.xml/route.ts`:

```ts
import { env } from "~/env";
import { getLastMod } from "~/lib/sitemap";
import { generateSitemaps as getArtistSitemaps } from "../api/sitemaps/artists/sitemap";
import { generateSitemaps as getEventSitemaps } from "../api/sitemaps/events/sitemap";
import { generateSitemaps as getVenueSitemaps } from "../api/sitemaps/venues/sitemap";

export const revalidate = 3600; // cache for 1 hour

function getFileName(id?: number) {
  if (id == null) {
    return "sitemap.xml";
  }
  /**
   * https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap#generating-multiple-sitemaps
   * In production, your generated sitemaps will be available at /.../sitemap/[id].xml. For example, /product/sitemap/1.xml.
   * In development, you can view the generated sitemap on /.../sitemap.xml/[id]. For example, /product/sitemap.xml/1. This difference is temporary and will follow the production format.
   * See the generateSitemaps API reference for more information.
   */
  return env.VERCEL_ENV === "development"
    ? `sitemap.xml/${id}`
    : `sitemap/${id}.xml`;
}
function getLoc(path: string, id?: number) {
  return `${env.URL}/api/sitemaps/${path}/${getFileName(id)}`;
}
function getSitemap(path: string, id?: number) {
  return /* XML */ `<sitemap><loc>${getLoc(path, id)}</loc><lastmod>${getLastMod()}</lastmod></sitemap>`;
}
function getSitemaps(ids: { id: number }[], path: string) {
  return ids.map(({ id }) => getSitemap(path, id)).join("");
}

export async function GET() {
  const xml = /* XML */ `<?xml version="1.0" encoding="UTF-8"?>
    <sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
      ${getSitemap("cities")}
      ${getSitemaps(await getEventSitemaps(), "events")}
      ${getSitemaps(await getArtistSitemaps(), "artists")}
      ${getSitemaps(await getVenueSitemaps(), "venues")}
    </sitemapindex>
  `;

  return new Response(xml, {
    headers: {
      "Content-Type": "application/xml",
    },
  });
}

```

**Reply by SemihOzden** [reply]
I have implemented same logic in my next.js project and It worked. Thanks in advance.

**Reply by jonathanlal** [reply]
legend thank you - elegant solution

**Comment by Sammeeey** [comment]
agreed, we need proper sitemap generation (incl. sitemap-index)

**Comment by pondorasti** [comment]
For anyone looking for a alternatives, I would recommend the server side utils from https://github.com/iamvishnusankar/next-sitemap?tab=readme-ov-file#server-side-sitemap-getserversidesitemap. You can plug them in into a route handler and it probably has all the features you would ever need.

**Comment by rtorrente** [comment]
I don't really understand the goal of generateSitemaps if it's just to generate the available url without indexing them. 
Paginated sitemap are growing indefinitly on a news site as post are created. It's just not possible to add new sitemap in google console because there is no alert like "New sitemap generated". Especially since, when using generateSitemap, the basic sitemap.xml file returns a 404, where indexes would be more appropriate. It seems more like a bug than a new feature at this level.

At least, **MetadataRoute.Sitemap** have to accept both url or sitemap link to allow dev to reference another sitemap file and generate an xml `<url></url>` or `<sitemap></sitemap>` accordingly.

**Reply by tomcru** [reply]
You create a Sitemap index that links to all your sitemaps - then you just add the index to google search console and it's able to find all of them. Like so: https://gamegator.net/sitemaps/index.xml (if that's what you mean)

**Reply by rtorrente** [reply]
Yes but it need to create a route and generate manually XML and we can't use he sitemap.ts at the root app/ directory

Because sitemap.ts only generate `<url>` XML tag and not `<sitemap>` XML tag

**Reply by heymartinadams** [reply]
Maybe this can help @rtorrente https://raddy.dev/blog/multiple-sitemaps-in-nextjs-app-router-sitemap-index/

**Comment by dsbrianwebster** [comment]
Sharing the approach we’ve been using:

We have a script that runs at the end of our build process. This script scans the app directory for all instances of `sitemap.ts`, resolves their URLs based on relative paths, and generates a single sitemap index at the root.

We like this approach because, once it’s in place, all we need to do is create sitemap.ts instances following Next.js [file conventions](https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap), with no additional work required. It’s also copy n' paste reusable across projects. Additionally, it even handles `sitemap.ts` files that use `generateSitemaps` and adds all the generated sitemaps to the index.

Here’s how you might integrate it into your package.json:

```json
  "scripts": {
    "build": "next build && pnpm sitemap:generate",
    "sitemap:generate": "dotenv -c -- npx tsx ./scripts/sitemap.ts",
  },
```

And here is the script... 

```ts
import { promises as fs } from 'fs';
import path from 'path';

import { SITE_URL } from '~/app/_lib/constants';

/**
 * 🗺️ Generates the main sitemap index file.
 *
 * This function creates a sitemap index file by discovering and aggregating
 * sitemap files generated across the project. It is temporary until Next.js
 * offers native support for sitemap indexing.
 */

const ROOT_DIR = process.cwd();
const APP_DIR = path.join(ROOT_DIR, 'app');

async function main() {
  try {
    const sitemaps = await findSitemapFiles(APP_DIR);
    const SITEMAP_INDEX_PATH = path.join(ROOT_DIR, 'public', '/sitemap.xml');

    const createSitemap = (url: string) => /* XML */ `
      <sitemap>
        <loc>${url}</loc>
      </sitemap>`;

    const createSitemapIndex = (urls: string[]) => /* XML */ `
      <?xml version="1.0" encoding="UTF-8"?>
      <sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
        ${urls.map(createSitemap).join('')}
      </sitemapindex>`;

    const xml = createSitemapIndex(sitemaps);
    await fs.mkdir(path.dirname(SITEMAP_INDEX_PATH), { recursive: true });
    await fs.writeFile(SITEMAP_INDEX_PATH, xml.trim());

    console.log(`Sitemap index generated with ${sitemaps.length} entries.`);
  } catch (error) {
    console.error('Error generating sitemap:', error);
  }
}

/**
 * 🔎 Recursively finds 'sitemap.ts' files and extracts sitemap URLs.
 *
 * Traverses directories from the specified start point, looking for 'sitemap.ts' files.
 * If a file contains a `generateSitemaps` function, the function is invoked
 * to retrieve sitemap URLs. If absent, a default sitemap URL is constructed.
 *
 * Known Limitations:
 * - While this does consider Next.js conventions for folders like route groups, do not place
 *   'sitemap.ts' under folders like route groups untill the following seperate issue is resolved:
 *    https://github.com/vercel/next.js/issues/68403
 *
 * @param {string} dir - The base directory to begin searching for sitemap files.
 * @returns {Promise<string[]>} A promise that resolves to an array of sitemap URLs.
 */
async function findSitemapFiles(dir: string): Promise<string[]> {
  let sitemaps: string[] = [];
  try {
    const files = await fs.readdir(dir, { withFileTypes: true });
    for (const file of files) {
      const fullPath = path.join(dir, file.name);
      if (file.isDirectory()) {
        // ingore private folders since all its child segments are opted out of routing
        // https://nextjs.org/docs/app/getting-started/project-structure#route-groups-and-private-folders
        if (!file.name.startsWith('_')) {
          const nestedSitemaps = await findSitemapFiles(fullPath);
          sitemaps = sitemaps.concat(nestedSitemaps);
        }
      } else if (file.isFile() && file.name === 'sitemap.ts') {
        const relativePath = path
          .relative(APP_DIR, dir)
          .split(path.sep)
          .filter(
            (segment) =>
              // Exclude specific Next.js folder conventions that should not be part of the resolved path
              // https://nextjs.org/docs/app/getting-started/project-structure#route-groups-and-private-folders
              // https://nextjs.org/docs/app/getting-started/project-structure#parallel-and-intercepted-routes
              // https://nextjs.org/docs/app/getting-started/project-structure#dynamic-routes
              !/^[(\[@]/.test(segment),
          )
          .join('/');
        const sitemapFile = await import(fullPath);
        if (sitemapFile.generateSitemaps) {
          const result = await sitemapFile.generateSitemaps();
          result.forEach((item: { id: number }) => {
            sitemaps.push(`${SITE_URL}/${relativePath}/sitemap/${item.id}.xml`);
          });
        } else {
          sitemaps.push(`${SITE_URL}/${relativePath}/sitemap.xml`);
        }
      }
    }
  } catch (error) {
    console.error('Error finding sitemaps:', error);
  }
  return sitemaps;
}

main();
```

**Reply by Zenith2198** [reply]
This is a somewhat clean solution that worked for me. When deploying a NextJS application to Vercel, I had to change a few things.

I was unable to get the SITE_URL from my envs, and had to hardcode it as the production domain. This was fine, because I only really need the sitemap to be correct on production.

I also changed the package.json scripts as follows:

```json
{
    "scripts": {
        "build": "next build",
        "postbuild": "npx tsx ./sitemap-index.ts"
    }
}
```

Loading dotenv in the script wasn't working for me, so I removed it. I also utilized postbuild to run the script.

You could consider making this into a standalone package. I think it would be very useful to people as an alternative to next-sitemap that integrates more cleanly with the built-in NextJS sitemap tools.

**Comment by Araon** [comment]
Ah yes, just what we needed—another 'half-baked feature' in Next.js that requires us to write 200 lines of custom scripts to make it actually useful. Love how we keep reinventing the wheel, but with extra steps. Can’t wait for the next ‘simplified’ feature that needs my entire engineering team to waste 1hr on.

**Reply by antoinerousseau** [reply]
What's the point of this comment? Is bitching really useful?

**Comment by programming-with-ia** [comment]
+1

**Comment by abel-officener** [comment]
+1

**Comment by Personal-Vijayji1137** [comment]
forgot every one comment follow this as this supports pagination also

step 1 :- create sitemap.xml folder and inside this create one file route.ts or route.js // here i have creted .ts


step 2 :- past code insite this file :-
```

import { NextRequest } from "next/server";
import ARTICLE_DB from "../server";

export async function GET(request: NextRequest) {
  const url = "https://www.narrato.in";

  const searchParams = request.nextUrl.searchParams;
  const page = parseInt(searchParams.get("page") || "1");

  const limit = 500;
  const startId = (page - 1) * limit;
  const endId = page * limit;
  try {
    const data = await ARTICLE_DB(`
      SELECT title, created_at FROM articles WHERE status = 'published' AND article_id BETWEEN ${startId} AND ${endId}
      ORDER BY created_at DESC
    `);

    const xmlItems = data
      .map(
        (item: any) => `
      <url>
        <loc>${url}/article/${item.title}</loc>
        <lastmod>${new Date(item.created_at).toISOString()}</lastmod>
        <changefreq>weekly</changefreq>
        <priority>1.0</priority>
      </url>`
      )
      .join("");
    const sitemapXml = `<?xml version="1.0" encoding="UTF-8"?>
    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
      <url>
        <loc>${url}</loc>
        <lastmod>${new Date().toISOString()}</lastmod>
        <changefreq>weekly</changefreq>
        <priority>1.0</priority>
      </url>
      <url>
        <loc>${url}/article</loc>
        <lastmod>${new Date().toISOString()}</lastmod>
        <changefreq>weekly</changefreq>
        <priority>0.8</priority>
      </url>
      ${xmlItems}
    </urlset>`;

    return new Response(sitemapXml, {
      headers: {
        "Content-Type": "application/xml",
      },
    });
  } catch (error) {
    const emptySitemap = `<?xml version="1.0" encoding="UTF-8"?>
    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"></urlset>`;

    return new Response(emptySitemap, {
      headers: {
        "Content-Type": "application/xml",
      },
    });
  }
}


```

**Comment by cozfuttu** [comment]
+1

**Comment by piotrkulpinski** [comment]
It would make it easier if at least the `sitemap.ts` with `generateSitemaps` wouldn't collide with a `sitemap.xml/route.ts` in the same route. 

The `sitemap.ts` doesn't seem to produce the `sitemap.xml` so I'm not sure why I'm getting this error: 
`⚠ Duplicate page detected. app/sitemap.ts and app/sitemap.xml/route.ts resolve to /sitemap.xml`

**Comment by programming-with-ia** [comment]
A better sitemap generator for Next.js App Router - full support for images, videos, hreflang alternates, sitemap indexes, custom endpoints, and real caching control with drop-in migration. [View on GitHub](https://github.com/programming-with-ia/better-next-sitemap)

**Comment by Forsect** [comment]
???

How is this still a thing 2 years later? You are literally forced to patch this half-baked feature if you want to utilize sitemaps in your app. Vercel?

**Reply by programming-with-ia** [reply]
watch this: https://github.com/programming-with-ia/better-next-sitemap
